### PR TITLE
REPO-2182 : Community: ALF-21852: setContentStream() returns the wrong document version when auto-versioning is turned on

### PR DESCRIPTION
**Pull request by Epurashu**
Separete into different new transactions writing content from creatingObjectId for setContentStream()
   Separate into different new transactions deleteing content from creatingObjectId for deleteContentStream()
   Separate into different new transations appending content from creatingObjectId for appendContentStream()
   Create JUnit CMISTest.testSetDeleteAppendContentStreamVersioning() testing that set/delete/apped calls return the new version.

### CODE REVIEW

File: src/main/java/org/alfresco/opencmis/AlfrescoCmisServiceImpl.java [hunk 1]
```diff
@@ -1526,15 +1527,32 @@ public void appendContentStream(String repositoryId, Holder<String> objectId, Ho
...
+               }
+                return null;
+            }}, false, true);
+
+       String objId = helper.doInTransaction(new RetryingTransactionCallback<String>()
+       {
+           public String execute() throws Throwable
+           {
+               return connector.createObjectId(nodeRef);
+       }}, false, true);
```
Comment (mmuller88) [inline src/main/java/org/alfresco/opencmis/AlfrescoCmisServiceImpl.java]: formation mistake?

File: src/main/java/org/alfresco/opencmis/AlfrescoCmisServiceImpl.java [hunk 2]
```diff
@@ -1569,25 +1587,41 @@ public void setContentStream(
...
+                return null;
+            }}, false, true);
 
-        objectId.setValue(connector.createObjectId(nodeRef));
+        String objId = helper.doInTransaction(new RetryingTransactionCallback<String>()
+        {
+            public String execute() throws Throwable
+            {
+                return connector.createObjectId(nodeRef);
+        }}, false, true);
```
Comment (mmuller88) [inline src/main/java/org/alfresco/opencmis/AlfrescoCmisServiceImpl.java]: here too

File: src/test/java/org/alfresco/opencmis/CMISTest.java [hunk 3]
```diff
@@ -3255,6 +3256,89 @@ private void checkDefs(List<TypeDefinitionWrapper> defs)
         withCmisService(callback, CmisVersion.CMIS_1_1);
     }
 
+    /**
+     * Test auto version behavior for setContentStream, deleteContentStream and appendContentStream according to REPO-2182.
```
Comment (mmuller88) [inline src/test/java/org/alfresco/opencmis/CMISTest.java]: Does it not has a Alf or MNT number

**Comment by derekhulley** [comment]
In some respects, this is a workaround for the fact that the versioning behaviour is only processed at the end of the transaction (I think that's what we're looking to work around in your fix).

File: src/main/java/org/alfresco/opencmis/AlfrescoCmisServiceImpl.java [hunk 4]
```diff
@@ -1526,15 +1527,32 @@ public void appendContentStream(String repositoryId, Holder<String> objectId, Ho
...
+                   connector.appendContent(info, contentStream, isLastChunk);
+               }
+               catch(IOException e)
+               {
+                    throw new ContentIOException("", e);
+               }
+                return null;
+            }}, false, true);
+
+       String objId = helper.doInTransaction(new RetryingTransactionCallback<String>()
```
Comment (derekhulley) [inline src/main/java/org/alfresco/opencmis/AlfrescoCmisServiceImpl.java]: This second transaction is independent of the first inner transaction; we've lost atomicity of the overall call.

File: src/main/java/org/alfresco/opencmis/AlfrescoCmisServiceImpl.java (hunk 2 above)
Comment (derekhulley) [inline src/main/java/org/alfresco/opencmis/AlfrescoCmisServiceImpl.java]: Again, atomicity is broken unless the second inner transaction is read-only.

**Comment by derekhulley** [comment]
That looks functional now.
Of course, there are two extra transactions per call, including the now-redundant transaction that wraps the method call.

An alternative approach would be to register a transaction listener to update the object ID is a post-transaction callback (afterCommit).  You would still need one of the read-only transactions but in a different phase.

On balance, I think the code you have is clearer than the alternative.

**Pull request merged**